File: socketd/src/main/java/org/noear/socketd/core/Session.java
package org.noear.socketd.core;

import java.io.Closeable;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.Map;
import java.util.function.Consumer;

/**
 * 会话
 *
 * @author noear
 * @since 2.0
 */
public interface Session extends Closeable {
    /**
     * 是否有效
     */
    boolean isValid();

    /**
     * 获取远程地址
     */
    InetSocketAddress getRemoteAddress() throws IOException;

    /**
     * 获取本地地址
     */
    InetSocketAddress getLocalAddress() throws IOException;

    /**
     * 获取握手信息
     */
    Handshaker getHandshaker();

    /**
     * 获取所有属性
     */
    Map<String,Object> getAttrMap();

    /**
     * 获取属性
     *
     * @param name 名字
     */
    <T> T getAttr(String name);

    /**
     * 设置属性
     *
     * @param name  名字
     * @param value 值
     */
    <T> void setAttr(String name, T value);

    /**
     * 获取会话Id
     */
    String getSessionId();

    /**
     * 发送 Ping
     */
    void sendPing() throws IOException;

    /**
     * 发送
     */
    void send(String topic, Entity content) throws IOException;

    /**
     * 发送并请求
     */
    default Entity sendAndRequest(String topic, Entity content) throws IOException {
        return sendAndRequest(topic, content, 3000);
    }

    /**
     * 发送并请求（限为一次答复）
     */
    Entity sendAndRequest(String topic, Entity content, long timeout) throws IOException;

    /**
     * 发送并订阅（不限答复次数）
     */
    void sendAndSubscribe(String topic, Entity content, Consumer<Entity> consumer) throws IOException;

    /**
     * 答复
     *
     * @param from 来源消息
     */
    void reply(Message from, Entity content) throws IOException;

    /**
     * 答复并结束
     *
     * @param from 来源消息
     */
    void replyEnd(Message from, Entity content) throws IOException;
}

File: python/socketd/socketd/core/Session.py
import abc
from typing import Union, Dict, Any, Callable, Awaitable, AsyncGenerator, Coroutine
from socket import gethostbyaddr
from socketd.core.Handshake import Handshake
from socketd.core.module.Message import Message
from socketd.core.module.Entity import Entity


class Session(abc.ABC):
    @abc.abstractmethod
    def is_valid(self) -> bool:
        ...

    @abc.abstractmethod
    def get_remote_address(self) -> gethostbyaddr:
        ...

    @abc.abstractmethod
    def get_local_address(self) -> gethostbyaddr:
        ...

    @abc.abstractmethod
    def get_handshake(self) -> Handshake:
        ...

    @abc.abstractmethod
    def get_attr_map(self) -> Dict[str, Any]:
        ...

    @abc.abstractmethod
    def get_attr(self, name: str) -> Union[None, Any]:
        ...

    @abc.abstractmethod
    def get_attr_or_default(self, name: str, default: Any) -> Any:
        ...

    @abc.abstractmethod
    def set_attr(self, name: str, value: Any) -> None:
        ...

    @abc.abstractmethod
    def get_session_id(self) -> str:
        ...

    @abc.abstractmethod
    def send_ping(self) -> Callable | Coroutine:
        ...

    @abc.abstractmethod
    async def send(self, event: str, content: Entity) -> None:
        ...

    @abc.abstractmethod
    async def send_and_request(self, event: str, content: Entity, timeout: int) -> Entity:
        ...

    @abc.abstractmethod
    async def send_stream_and_request(self, event: str, content: Entity,
                                      consumer: Callable[[Entity], Awaitable[Any]] | Coroutine[Entity, Any, None],
                                      timeout: int):
        """
        发送流和请求的抽象方法。

        Args:
            event (str): 事件名称。
            content (Entity): 内容实体。
            consumer (Callable[[Entity], Awaitable[Any]] | Coroutine[Entity, Awaitable[Any]]): 消费函数或协程，用来处理内容实体并返回异步结果。
            timeout (int): 超时时间。

        Returns:
            Awaitable[Any]: 消费函数或协程的异步结果。
        """
        ...

    @abc.abstractmethod
    async def send_and_subscribe(self, topic: str, content: Entity, consumer: Callable[[Entity], Any],
                                 timeout: int) -> None:
        ...

    @abc.abstractmethod
    def reply(self, from_msg: Message, content: Entity) -> None:
        ...

    @abc.abstractmethod
    def reply_end(self, from_msg: Message, content: Entity) -> None:
        ...

    @abc.abstractmethod
    def close(self):
        ...
